internal/rpc/clients.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpc
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"open-match.dev/open-match/internal/config"
+)
+
+var (
+	clientLogger = logrus.WithFields(logrus.Fields{
+		"app":       "openmatch",
+		"component": "client",
+	})
+)
+
+// ClientParams contains the connection parameters to connect to an Open Match service.
+type ClientParams struct {
+	Hostname          string
+	Port              int
+	PublicCertificate []byte
+	PrivateKeyData    []byte
+}
+
+func (p *ClientParams) usingTLS() bool {
+	return len(p.PublicCertificate) > 0
+}
+
+// GRPCClientFromConfig creates a gRPC client connection from a configuration.
+func GRPCClientFromConfig(cfg config.View, prefix string) (*grpc.ClientConn, error) {
+	hostname := cfg.GetString(prefix + ".hostname")
+	portNumber := cfg.GetInt(prefix + ".grpcport")
+
+	// If TLS support is not enabled in the config, return an insecure grpc client.
+	if !cfg.GetBool("tls.enabled") {
+		return GRPCClientFromParams(&ClientParams{
+			Hostname: hostname,
+			Port:     portNumber,
+		})
+	}
+
+	// If TLS support is enabled, read and return a grpc TLS client.
+	publicKeyPath := cfg.GetString("tls.publicKeyPath")
+
+	publicKeyBytes, err := ioutil.ReadFile(publicKeyPath)
+	if err != nil {
+		clientLogger.WithError(err).Fatalf("cannot read public key from file %s", publicKeyPath)
+		return nil, err
+	}
+
+	return GRPCClientFromParams(&ClientParams{
+		Hostname:          hostname,
+		Port:              portNumber,
+		PublicCertificate: publicKeyBytes,
+	})
+}
+
+// GRPCClientFromParams creates a gRPC client connection from the parameters.
+func GRPCClientFromParams(params *ClientParams) (*grpc.ClientConn, error) {
+	address := fmt.Sprintf("%s:%d", params.Hostname, params.Port)
+	if params.usingTLS() {
+		creds, err := clientCredentialsFromFileData(params.PublicCertificate, address)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return grpc.Dial(address, grpc.WithTransportCredentials(creds))
+	}
+	return grpc.Dial(address, grpc.WithInsecure())
+}
+
+// HTTPClientFromConfig creates a HTTP client from from a configuration.
+func HTTPClientFromConfig(cfg config.View, prefix string) (*http.Client, string, error) {
+	hostname := cfg.GetString(prefix + ".hostname")
```

Comment:
This code would be cleaner if you make a ClientParams variable at the top, and fill it out throughout the function.  For the whole if statement about tls being enabled, you can invert it and only run it if it is enabled.  Within the if statement, add the fields tls parameters to the existing ClientParams variable.

Reply:
Will fix

Reply:
Please only resolve once you have the fix in the PR.